### ISSUE TITLE
Kotlin/JS IR backend breaks when using BuildKonfig's `exposeObjectWithName` option

### ISSUE THREAD

**Issue by chris-hatton** [yshrsmz/BuildKonfig#54]
As soon as I enable the Config file generation, even with no fields defined, the task `compileProductionExecutableKotlinBrowser` yields error:
```
java.lang.AssertionError: Unexpected field without property myapp.MyConfig.$stable
	at org.jetbrains.kotlin.ir.backend.js.export.ExportModelGenerator.exportClass(ExportModelGenerator.kt:177)
	at org.jetbrains.kotlin.ir.backend.js.export.ExportModelGenerator.exportDeclaration(ExportModelGenerator.kt:58)
	at org.jetbrains.kotlin.ir.backend.js.export.ExportModelGenerator.generateExport(ExportModelGenerator.kt:34)
	at org.jetbrains.kotlin.ir.backend.js.export.ExportModelGenerator.generateExport(ExportModelGenerator.kt:47)
	at org.jetbrains.kotlin.ir.backend.js.transformers.irToJs.IrModuleToJsTransformer.generateModule(IrModuleToJsTransformer.kt:46)
	at org.jetbrains.kotlin.ir.backend.js.CompilerKt.compile(compiler.kt:123)
	at org.jetbrains.kotlin.ir.backend.js.CompilerKt.compile$default(compiler.kt:36)
	at org.jetbrains.kotlin.cli.js.K2JsIrCompiler.doExecute(K2JsIrCompiler.kt:256)
	at org.jetbrains.kotlin.cli.js.K2JSCompiler.doExecute(K2JSCompiler.java:182)
	at org.jetbrains.kotlin.cli.js.K2JSCompiler.doExecute(K2JSCompiler.java:75)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:90)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:44)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:98)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl.compile(CompileServiceImpl.kt:1575)
	at jdk.internal.reflect.GeneratedMethodAccessor105.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:359)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:200)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:197)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.rmi/sun.rmi.transport.Transport.serviceCall(Transport.java:196)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:562)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:796)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:677)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:676)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by chris-hatton** [comment]
Reproduced with Gradle versions `7.1.1` and `7.2`.

**Comment by yshrsmz** [comment]
can you give me a reproducible repo or gradle config file?

**Comment by chris-hatton** [comment]
Yes, I will provide a minimal reproduction project ASAP.

**Comment by chris-hatton** [comment]
I've created a minimal reproduction project; but the process of creating it revealed that this issue is probably the side effect of a bug between Kotlin/JS and the JetBrains Compose compiler.  With the Compose compiler disabled, BuildKonfig works correctly, when it's enabled, the stated assertion error is thrown.

I raised this as an issue on:
- [YouTrack](https://youtrack.jetbrains.com/issue/KT-48366) (minimal reproduction project attached here)
- [Kotlin Slack](https://kotlinlang.slack.com/archives/CJLTWPH7S/p1629723189384700)

**Comment by yshrsmz** [comment]
Thanks!

**Comment by G00fY2** [comment]
I just added the plugin to a KotlinJS Compose project and everything is working fine so far: https://github.com/G00fY2/compose-github.io/commit/08a3f442600008fa833c7f77465550b3960950b0

But I am using compose 1.0.0-alpha4-build328 with Kotlin 1.5.30.

**Comment by chris-hatton** [comment]
@G00fY2 Just adding the plugin alone is not enough to reproduce the error, you have to add a field to your Config file so that it gets generated.  If this is working for you; would you kindly share your config and/or also try to compile the reproduction project in the YouTrack issue so that we can learn what the difference is?  Thanks.

**Comment by chris-hatton** [comment]
@G00fY2 Thank you for your working example; using this I was able to narrow the issue down further, to use of BuildKonfigs `exposeObjectWithName` (see [documentation](https://github.com/yshrsmz/BuildKonfig#kotlin-dsl)).  Whenever this is set the problem occurs; it looks like the relevant factor is the generated output then having `public` instead of `internal` visibility.

**Comment by G00fY2** [comment]
@chris-hatton As I said, it's working in my project (I actually switched to `om.github.gmazzo.buildconfig` since I only need KotlinJs and want the fields to be _const_).

I tried your sample project and the issue seems to be the `exposeObjectWithName` setting. I didn't use that. As soon as you remove this, your project compiles. However not sure whats the root cause.

**Comment by chris-hatton** [comment]
Yeah, I recreated the issue in your project by setting `exposeObjectWithName` (and adding requisite ` "-Xopt-in=kotlin.RequiresOptIn"` JS Compiler arg) so it seem to be a Kotlin/JS issue with generated code being public.

**Comment by yshrsmz** [comment]
@chris-hatton @G00fY2

Hey, can anyone confirm this?

https://github.com/JetBrains/compose-jb/issues/1104 says that it should be fixed with 1.0.1-rc2 and kotlin 1.6.10

**Comment by yshrsmz** [comment]
closing for now.
feel free to reopen if this is still an issue.
